pkg/apis/projectcalico/v3/nodestatus.go

```diff
@@ -0,0 +1,201 @@
+	// last updated. It is represented in RFC3339 form and is in UTC.
+	// +nullable
+	LastUpdated metav1.Time `json:"lastUpdated,omitempty"`
+
+	// AdditionalInfo is a a human-readable description of the status of last update.
+	AdditionalInfo string `json:"additionalInfo,omitempty"`
+
+	// Agent holds agent status on the node.
+	Agent CalicoNodeAgentStatus `json:"agent,omitempty"`
+
+	// BGP holds node BGP status.
+	BGP CalicoNodeBGPStatus `json:"bgp,omitempty"`
+
+	// Route represents routes on the node.
+	Route CalicoNodeRouteStatus `json:"route,omitempty"`
+}
+
+// CalicoNodeAgentStatus defines the observed state of agent status on the node.
+type CalicoNodeAgentStatus struct {
+	// Bird4 represents the latest observed status of bird4.
+	Bird4 CalicoNodeBirdStatus `json:"bird4,omitempty"`
+
+	// Bird6 represents the latest observed status of bird6.
+	Bird6 CalicoNodeBirdStatus `json:"bird6,omitempty"`
+}
+
+// CalicoNodeBGPStatus defines the observed state of BGP status on the node.
+type CalicoNodeBGPStatus struct {
+	// The total number of established bgp sessions.
+	NumEstablished int `json:"numEstablished,omitempty"`
+
+	// The total number of non-established bgp sessions.
+	NumNotEstablished int `json:"numNotEstablished,omitempty"`
+
+	// V4Peers represents IPv4 BGP peers status on the node.
+	V4Peers []CalicoNodePeer `json:"v4Peers,omitempty"`
+
+	// V6Peers represents IPv6 BGP peers status on the node.
+	V6Peers []CalicoNodePeer `json:"v6Peers,omitempty"`
+}
+
+// CalicoNodeRouteStatus defines the observed state of routes status on the node.
+type CalicoNodeRouteStatus struct {
+	// V4 represents IPv4 routes on the node.
+	V4 []CalicoNodeRoute `json:"bird4,omitempty"`
+
+	// V6 represents IPv6 routes on the node.
+	V6 []CalicoNodeRoute `json:"bird6,omitempty"`
+}
+
+// CalicoNodeBirdStatus defines the observed state of bird.
+type CalicoNodeBirdStatus struct {
+	// Ready indicates if bird status is ready.
+	Ready bool `json:"ready,omitempty"`
+
+	// Bird version.
+	Version string `json:"version,omitempty"`
+
+	// Route ID used by bird.
+	RouteID string `json:"routeID,omitempty"`
+
+	// ServerTime holds the value of serverTime from birdctl output.
+	ServerTime string `json:"serverTime,omitempty"`
+
+	// LastBootTime holds the value of lastBootTime from birdctl output.
+	LastBootTime string `json:"lastBootTime,omitempty"`
+
+	// LastReconfigTime holds the value of lastReconfigTime from birdctl output.
+	LastReconfigTime string `json:"lastReconfigTime,omitempty"`
+}
+
+// CalicoNodePeer contains the status of BGP peers on the node.
+type CalicoNodePeer struct {
+	// IP address of the peer whose condition we are reporting.
+	// If port number is given, format should be `[<IPv6>]:port` or `<IPv4>:<port>` for IPv4.
+	// If optional port number is not set, and this peer IP and ASNumber belongs to a calico/node
+	// with ListenPort set in BGPConfiguration, then we use that port to peer.
+	PeerIP string `json:"peerIP,omitempty"`
```

Comment:
I don't think I understand how the port comes into play on this field - is there always a `:port` set? Or is that only included if using a non-standard port?

I think we can simplify this entire thing down to something like this:

```go
// IP address and port of this BGP peer, separated by a colon. 
PeerIP string `json"peerIP,omitempty"`
```

Reply:
I copied the field from BGPPeer but you are right, in our case all the information we get is the IP address.

Reply:
Are you saying, @caseydavenport , that the port must **always** be present?  If so, fine.  If not, there can be ambiguity in some IPv6 cases if `[` `]` are not used.

Reply:
Yeah we'd need to properly format the address if we do that. The reason to always specify the port is just to make the API more consumable - it's confusing when fields have data which is formatted conditionally based on another field, or in this case something that isn't even in this API strucutre.

We could either have separate IP / port fields, or just always include the port in this field. I am probably OK with either, although separate fields is more explicit.

Reply:
@song-jiang what was the resolution on this one - are we just not including the port at all?